src/main/java/com/mongodb/kafka/connect/source/topic/mapping/DefaultTopicMapper.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.kafka.connect.source.topic.mapping;
+
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.TOPIC_NAMESPACE_MAP_CONFIG;
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.TOPIC_PREFIX_CONFIG;
+import static com.mongodb.kafka.connect.source.MongoSourceConfig.TOPIC_SUFFIX_CONFIG;
+import static com.mongodb.kafka.connect.util.ConfigHelper.documentFromString;
+import static java.lang.String.format;
+
+import org.bson.BsonDocument;
+import org.bson.BsonString;
+import org.bson.Document;
+
+import com.mongodb.kafka.connect.source.MongoSourceConfig;
+
+public class DefaultTopicMapper implements TopicMapper {
+
+  private static final String NS_KEY = "ns";
+  private static final String DB_KEY = "db";
+  private static final String COLL_KEY = "coll";
+  private static final String ALL = "*";
+
+  private String prefix;
+  private String suffix;
+  private Document topicNamespaceMap;
+
+  @Override
+  public void configure(final MongoSourceConfig config) {
+    String prefix = config.getString(TOPIC_PREFIX_CONFIG).trim();
+    String suffix = config.getString(TOPIC_SUFFIX_CONFIG).trim();
+
+    this.prefix = prefix.isEmpty() ? "" : format("%s.", prefix);
+    this.suffix = suffix.isEmpty() ? "" : format(".%s", suffix);
+    this.topicNamespaceMap =
+        documentFromString(config.getString(TOPIC_NAMESPACE_MAP_CONFIG)).orElse(new Document());
+  }
+
+  @Override
+  public String getTopic(final BsonDocument changeStreamDocument) {
+    BsonDocument namespaceDocument = changeStreamDocument.getDocument(NS_KEY, new BsonDocument());
+    String dbName = namespaceDocument.getString(DB_KEY, new BsonString("")).getValue();
+    String collName = namespaceDocument.getString(COLL_KEY, new BsonString("")).getValue();
+    String topicName = collName.isEmpty() ? dbName : format("%s.%s", dbName, collName);
+
+    if (topicNamespaceMap.containsKey(topicName)) {
+      topicName = topicNamespaceMap.get(topicName).toString();
+    } else if (topicNamespaceMap.containsKey(dbName) && !collName.isEmpty()) {
+      topicName = format("%s.%s", topicNamespaceMap.get(dbName), collName);
+    } else if (topicNamespaceMap.containsKey(ALL)) {
+      topicName = topicNamespaceMap.getString(ALL);
```

Comment:
Agreed - I've renamed it to namespace and then the final output is the topic.

The topicNamespaceMap (namespace:topic) is unfortunate, however as the configuration all starts with "topic" eg. "topic.prefix" then "topic.namespace.map" seemed the best choice for the configuration.  Happy to take an alternative name though.

Reply:
I still see a single variable used as both namespace and topic;  was thinking about something like the following:
```
String namespace = collName.isEmpty() ? dbName : format("%s.%s", dbName, collName);
String topic = namespace;
if (topicNamespaceMap.containsKey(namespace)) {
  topic = topicNamespaceMap.get(namespace).toString();
} else if (topicNamespaceMap.containsKey(dbName) && !collName.isEmpty()) {
  topic = format("%s.%s", topicNamespaceMap.get(dbName), collName);
} else if (topicNamespaceMap.containsKey(ALL)) {
  topic = topicNamespaceMap.getString(ALL);
}

return format("%s%s%s", prefix, topic, suffix);
```

However, please feel free to just resolve this conversation if you feel like it.

P.S. As a side note, I see us searching `topicNamespaceMap` twice for each key and using `format` where concatenation is sufficient. These are all trivial things that do not affect readability much (arguably, concatenation improves it in this case), but have significantly change the performance of the `getTopic` method, which is executed for every document. I agree with the approach of not sacrificing readability unless there are performance issues, but I also not sure I see a reason to produce less efficient code just because no one has yet reported a performance issue.

Reply:
See its not the actual topic until the prefix / suffix has been added. So I'm inclined to leave as is.

Will look at the performance issue separately

Reply:
I've rewritten the code. Adding a cache and handling invalid data.

Reply:
Thanks, looks good! There is only only one thing left that I am not sure I understand (asked in a separate comment).